changeset/manager.go

```diff
@@ -0,0 +1,415 @@
+// commits are not applied more than once, and to avoid conflicts on the change
+// object itself.
+type Manager interface {
+	// Change creates a new change against the latest configuration, adding it
+	// to the set of pending changes for that configuration
+	Change(change ChangeFn) error
+
+	// GetPendingChanges gets the latest uncommitted changes
+	GetPendingChanges() (int, proto.Message, proto.Message, error)
+
+	// Commit commits the specified ChangeSet, transforming the configuration on
+	// which they are based into a new configuration, and storing that new
+	// configuration as a next versions. Ensures that changes are applied as a
+	// batch, are not applied more than once, and that new changes are not
+	// started while a commit is underway
+	Commit(version int, apply ApplyFn) error
+}
+
+// NewManager creates a new change list Manager
+func NewManager(opts ManagerOptions) Manager {
+	if opts == nil {
+		panic("opts must not be nil")
+	}
+
+	if opts.GetConfigKey() == "" {
+		panic("opts.ConfigKey must be set")
+	}
+
+	if opts.GetKV() == nil {
+		panic("opts.KV must be set")
+	}
+
+	if opts.GetConfigType() == nil {
+		panic("opts.ConfigType must be set")
+	}
+
+	if opts.GetChangesType() == nil {
+		panic("opts.ChangesType must be set")
+	}
+
+	logger := opts.GetLogger()
+	if logger == nil {
+		logger = xlog.NullLogger
+	}
+
+	return &manager{
+		key:         opts.GetConfigKey(),
+		kv:          opts.GetKV(),
+		configType:  proto.Clone(opts.GetConfigType()),
+		changesType: proto.Clone(opts.GetChangesType()),
+		log:         logger,
+	}
+}
+
+type manager struct {
+	key         string
+	kv          kv.Store
+	configType  proto.Message
+	changesType proto.Message
+	log         xlog.Logger
+}
+
+func (m manager) Change(change ChangeFn) error {
+	for {
+		// Retrieve the current configuration, creating an empty config if one does
+		// not exist
+		config := proto.Clone(m.configType)
+		configVersion, err := m.getOrCreate(m.key, config)
+		if err != nil {
+			return err
+		}
+
+		// Retrieve the changes for the current configuration, creating an empty
+		// change set if one does not exist
+		changeset := &changesetpb.ChangeSet{
+			ForVersion: int32(configVersion),
+			State:      changesetpb.ChangeSetState_OPEN,
+		}
+
+		csKey := fmtChangeSetKey(m.key, configVersion)
+		csVersion, err := m.getOrCreate(csKey, changeset)
+		if err != nil {
+			return err
+		}
+
+		// Only allow changes to an open change list
+		if changeset.State != changesetpb.ChangeSetState_OPEN {
+			return ErrCommitInProgress
+		}
+
+		// Apply the new changes...
+		changes := proto.Clone(m.changesType)
+		if err := proto.Unmarshal(changeset.Changes, changes); err != nil {
+			return err
+		}
+
+		if err := change(config, changes); err != nil {
+			return err
+		}
+
+		changeBytes, err := proto.Marshal(changes)
+		if err != nil {
+			return err
+		}
+
+		// ...and update the stored changes
+		changeset.Changes = changeBytes
```

Comment:
I see, so for a given `configType`, in order to make multiple changes, the underlying `changesType` must allow for the caller to inspect the previous changes inside the saved `changesType` and mutate it such that both all that came before it and its own changes are captured in it by the mutations it makes itself to instance of the `changesType` passed to it.  

Ok I get it now.  This means we won't natively be able to capture the length of modifications made nor stringify each one individually.  I think this makes sense but because we don't explicitly store them separately it will reduce the capability somewhat of tooling that only has access to a concrete `Manager`.